components/match2/wikis/clashroyale/match_group_input_custom.lua

```diff
@@ -5,3 +5,597 @@
+
+		--process input
+		if opponent.type == Opponent.team or opponent.type == Opponent.solo or
+			opponent.type == Opponent.duo or opponent.type == Opponent.literal then
+
+			opponent = CustomMatchGroupInput.processOpponent(opponent, match.timestamp)
+		else
+			error('Unsupported Opponent Type')
+		end
+
+		-- Retrieve icon for team
+		if opponent.type == Opponent.team then
+			opponent.icon, opponent.icondark = CustomMatchGroupInput.getIcon(opponent.template)
+			match.hasTeamOpponent = true
+		end
+
+		--set initial opponent sumscore
+		opponent.sumscore = 0
+
+		match['opponent' .. opponentIndex] = opponent
+
+		opponentIndex = opponentIndex + 1
+		opponent = match['opponent' .. opponentIndex]
+	end
+
+	return match
+end
+
+function CustomMatchGroupInput.processOpponent(record, timestamp)
+	local opponent = Opponent.readOpponentArgs(record)
+		or Opponent.blank()
+
+	local teamTemplateDate = timestamp
+	-- If date is epoch, resolve using tournament dates instead
+	-- Epoch indicates that the match is missing a date
+	-- In order to get correct child team template, we will use an approximately date and not 1970-01-01
+	if teamTemplateDate == DateExt.epochZero then
+		teamTemplateDate = Variables.varDefaultMulti(
+			'tournament_enddate',
+			'tournament_startdate',
+			TODAY
+		)
+	end
+
+	Opponent.resolve(opponent, teamTemplateDate)
+
+	MatchGroupInput.mergeRecordWithOpponent(record, opponent)
+
+	if record.type == Opponent.team then
+		record.match2players = CustomMatchGroupInput._readTeamPlayers(record, record.players)
+	end
+
+	return record
+end
+
+function CustomMatchGroupInput._readTeamPlayers(opponent, playerData)
+	local players = CustomMatchGroupInput._getManuallyEnteredPlayers(playerData)
+	opponent.players = nil
+
+	if Table.isEmpty(players) then
+		players = CustomMatchGroupInput._getPlayersFromVariables(opponent.name)
+	end
+
+	return players
+end
+
+function CustomMatchGroupInput._getManuallyEnteredPlayers(playerData)
+	local players = {}
+	playerData = Json.parseIfString(playerData) or {}
+
+	for prefix, displayName in Table.iter.pairsByPrefix(playerData, 'p') do
+		local name = mw.ext.TeamLiquidIntegration.resolve_redirect(Logic.emptyOr(
+			playerData[prefix .. 'link'],
+			displayName
+		))
+
+		table.insert(players, {
+			name = name,
+			displayname = displayName,
+			flag = Flags.CountryName(playerData[prefix .. 'flag']),
+		})
+	end
+
+	return players
+end
+
+function CustomMatchGroupInput._getPlayersFromVariables(teamName)
+	local players = {}
+
+	local playerIndex = 1
+	local prefix = teamName .. '_p' .. playerIndex
+	local playerName = Variables.varDefault(prefix)
+	while String.isNotEmpty(playerName) do
+		table.insert(players, {
+			name = playerName,
+			displayname = Variables.varDefault(prefix .. 'dn'),
+			flag = Flags.CountryName(Variables.varDefault(prefix .. 'flag')),
+		})
+
+		playerIndex = playerIndex + 1
+		prefix = teamName .. '_p' .. playerIndex
+		playerName = Variables.varDefault(prefix)
+	end
```

Comment:
PairsByPrefix

Reply:
how?
we are loking at wiki vars not a table

Reply:
Missed that. 

```suggestion
	local playerIndex = 1
	while true do
		local prefix = teamName .. '_p' .. playerIndex
		local playerName = Variables.varDefault(prefix)
		if String.isEmpty(playerName) then
			break
		end
		table.insert(players, {
			name = playerName,
			displayname = Variables.varDefault(prefix .. 'dn'),
			flag = Flags.CountryName(Variables.varDefault(prefix .. 'flag')),
		})
		playerIndex = playerIndex + 1
	end
```
No repetition of logic with this variant